// @zxcvbn-ts/matcher-pwned/dist/zxcvbn-ts.js
this.zxcvbnts = this.zxcvbnts || {};
this.zxcvbnts["matcher-pwned"] = (function (exports, core) {
    'use strict';

    const textEncode = text => {
      try {
        return new TextEncoder().encode(text);
      } catch (error) {
        throw new Error(`No encoder found, ${error}`);
      }
    };
    const digestMessage = async message => {
      const data = textEncode(message);
      const hashBuffer = await crypto.subtle.digest('SHA-1', data);
      const hashArray = Array.from(new Uint8Array(hashBuffer));
      return hashArray.map(b => b.toString(16).padStart(2, '0')).join('').toUpperCase();
    };
    const pwnedUrl = 'https://api.pwnedpasswords.com/range/';
    const defaultNetworkErrorHandler = _error => {
      return false;
    };
    var haveIBeenPwned = async (password, {
      universalFetch,
      url = pwnedUrl,
      networkErrorHandler = defaultNetworkErrorHandler
    }) => {
      // we don't need to check passwords with a length smaller 2 which can happen for repeat matcher or if we don't have a fetch function
      if (!universalFetch || password.length < 2) {
        return null;
      }
      const passwordHash = (await digestMessage(password)).toUpperCase();
      const range = passwordHash.slice(0, 5);
      const suffix = passwordHash.slice(5);
      const response = await universalFetch(`${url}${range}`, {
        method: 'GET',
        headers: {
          'Add-Padding': 'true'
        }
      }).catch(error => {
        return networkErrorHandler(error);
      });
      if (typeof response === 'boolean') {
        return false;
      }
      if (response.status >= 400) {
        return networkErrorHandler(response);
      }
      const result = await response.text();
      const resultArray = result.split('\r\n');
      return resultArray.find(entry => {
        const passwordHasPart = entry.split(':')[0];
        return passwordHasPart === suffix;
      });
    };

    const matchingFactory = (universalFetch, {
      url,
      networkErrorHandler
    }) => {
      class MatchPwned extends core.MatcherBaseClass {
        async match({
          password
        }) {
          const matches = [];
          const pwned = await haveIBeenPwned(password, {
            universalFetch,
            url,
            networkErrorHandler
          });
          if (pwned) {
            matches.push({
              pattern: 'pwned',
              pwnedAmount: parseInt(pwned.split(':')[1], 10),
              i: 0,
              j: password.length - 1,
              token: password
            });
          }
          return matches;
        }
      }
      return MatchPwned;
    };

    // TODO make some more appropriated guesses logic?
    var scoring = () => {
      return 1;
    };

    const pwnedFeedback = options => {
      return {
        warning: options.translations.warnings.pwned,
        suggestions: [options.translations.suggestions.pwned]
      };
    };

    const matcherPwnedFactory = (universalFetch, config = {}) => {
      return {
        Matching: matchingFactory(universalFetch, config),
        feedback: pwnedFeedback,
        scoring
      };
    };

    exports.haveIBeenPwned = haveIBeenPwned;
    exports.matcherPwnedFactory = matcherPwnedFactory;

    return exports;

})({}, core);
